store/pg_engine.go

```diff
@@ -308,55 +319,47 @@ func migrate(ctx context.Context, d dbdriver.Driver, curVer *semver.Version, mod
 			},
 			stmt,
 		); err != nil {
-			return semver.Version{}, fmt.Errorf("failed to migrate initial schema version %q, error: %v", latestSchemaPath, err)
+			return fmt.Errorf("failed to migrate initial schema version %q, error: %v", latestSchemaPath, err)
 		}
 		l.Info(fmt.Sprintf("Completed database initial migration with version %s.", cutoffSchemaVersion))
-		return cutoffSchemaVersion, nil
-	}
-
-	// Apply migrations
-	retVersion := *curVer
-
-	releaseModes := []common.ReleaseMode{common.ReleaseModeRelease}
-	if mode == common.ReleaseModeDev {
-		releaseModes = append(releaseModes, common.ReleaseModeDev)
-	}
-
-	for _, releaseMode := range releaseModes {
-		names, err := fs.Glob(migrationFS, fmt.Sprintf("migration/%s/*", releaseMode))
+	} else {
+		// Apply migrations if needed.
+		retVersion := *curVer
+		names, err := fs.Glob(migrationFS, fmt.Sprintf("migration/%s/*", common.ReleaseModeRelease))
 		if err != nil {
-			return semver.Version{}, err
+			return err
 		}
 
 		minorVersions, messages, err := getMinorMigrationVersions(names, cutoffSchemaVersion, *curVer)
 		if err != nil {
-			return semver.Version{}, err
+			return err
 		}
 		for _, message := range messages {
 			l.Info(message)
 		}
-		if len(minorVersions) == 0 {
-			l.Info(fmt.Sprintf("Skip minor version migration in mode %s. No new version.", releaseMode))
-			continue
-		}
 
 		for _, minorVersion := range minorVersions {
 			l.Info(fmt.Sprintf("Starting minor version migration cycle from %s ...", minorVersion))
-			names, err := fs.Glob(migrationFS, fmt.Sprintf("migration/%s/%d.%d/*.sql", releaseMode, minorVersion.Major, minorVersion.Minor))
+			names, err := fs.Glob(migrationFS, fmt.Sprintf("migration/%s/%d.%d/*.sql", common.ReleaseModeRelease, minorVersion.Major, minorVersion.Minor))
 			if err != nil {
-				return semver.Version{}, err
+				return err
 			}
 			patchVersions, err := getPatchVersions(minorVersion, cutoffSchemaVersion, *curVer, names)
 			if err != nil {
-				return semver.Version{}, err
+				return err
 			}
 
 			for _, pv := range patchVersions {
-				l.Info(fmt.Sprintf("Migrating %s...", pv.version))
 				buf, err := fs.ReadFile(migrationFS, pv.filename)
 				if err != nil {
-					return semver.Version{}, fmt.Errorf("failed to read migration file %q, error %w", pv.filename, err)
+					return fmt.Errorf("failed to read migration file %q, error %w", pv.filename, err)
+				}
+				// This happens when a migration file is moved from dev to release and we should not reapply the migration.
+				if mode == common.ReleaseModeDev && migrationExists(string(buf), histories) {
```

Comment:
Example:
before
  release 1.2
  dev: 123.sql, something else
after
  release 1.3 with 123.sql
  dev: something else

When dev starts, it will try to apply version 1.3 including 123.sql. If we don't skip, the same statement will be re-apply and most likely to fail.

Reply:
You are right. I misread. BTW, I would change to use Prod instead of Release. Currently, "Release is also a Release" which is confusing.

Reply:
SG. I will rename release to prod later.